clisops/utils/dataset_utils.py

```diff
@@ -368,6 +368,556 @@ def adjust_date_to_calendar(da, date, direction="backwards"):
+
+    lons_crnr = np.empty((ds[lon].shape[0] + 1, ds[lon].shape[1] + 1))
+    lons_crnr[:] = np.nan
+    lats_crnr = np.empty((ds[lat].shape[0] + 1, ds[lat].shape[1] + 1))
+    lats_crnr[:] = np.nan
+
+    # -------- Calculating corners --------- #
+
+    # Loop through all grid points except at the boundaries
+    for ilat in range(1, ds[lon].shape[0]):
+        for ilon in range(1, ds[lon].shape[1]):
+            # SW corner for each lat/lon index is calculated
+            lons_crnr[ilat, ilon] = (
+                ds[lon][ilat - 1, ilon - 1]
+                + ds[lon][ilat, ilon - 1]
+                + ds[lon][ilat - 1, ilon]
+                + ds[lon][ilat, ilon]
+            ) / 4.0
+            lats_crnr[ilat, ilon] = (
+                ds[lat][ilat - 1, ilon - 1]
+                + ds[lat][ilat, ilon - 1]
+                + ds[lat][ilat - 1, ilon]
+                + ds[lat][ilat, ilon]
+            ) / 4.0
+
+    # Grid points at boundaries
+    lons_crnr[0, :] = lons_crnr[1, :] - (lons_crnr[2, :] - lons_crnr[1, :])
+    lons_crnr[-1, :] = lons_crnr[-2, :] + (lons_crnr[-2, :] - lons_crnr[-3, :])
+    lons_crnr[:, 0] = lons_crnr[:, 1] + (lons_crnr[:, 1] - lons_crnr[:, 2])
+    lons_crnr[:, -1] = lons_crnr[:, -2] + (lons_crnr[:, -2] - lons_crnr[:, -3])
+
+    lats_crnr[0, :] = lats_crnr[1, :] - (lats_crnr[2, :] - lats_crnr[1, :])
+    lats_crnr[-1, :] = lats_crnr[-2, :] + (lats_crnr[-2, :] - lats_crnr[-3, :])
+    lats_crnr[:, 0] = lats_crnr[:, 1] - (lats_crnr[:, 1] - lats_crnr[:, 2])
+    lats_crnr[:, -1] = lats_crnr[:, -2] + (lats_crnr[:, -2] - lats_crnr[:, -3])
+
+    # ------------ DONE ------------- #
+
+    # Fill in counterclockwise and rearrange
+    count = 0
+    for ilat in range(ds[lon].shape[0]):
+        for ilon in range(ds[lon].shape[1]):
+            lons_cor[count] = lons_crnr[ilat, ilon]
+            lons_cor[count + 1] = lons_crnr[ilat, ilon + 1]
+            lons_cor[count + 2] = lons_crnr[ilat + 1, ilon + 1]
+            lons_cor[count + 3] = lons_crnr[ilat + 1, ilon]
+
+            lats_cor[count] = lats_crnr[ilat, ilon]
+            lats_cor[count + 1] = lats_crnr[ilat, ilon + 1]
+            lats_cor[count + 2] = lats_crnr[ilat + 1, ilon + 1]
+            lats_cor[count + 3] = lats_crnr[ilat + 1, ilon]
+
+            count += 4
+
+    lon_bnds = lons_cor.reshape(nlat, nlon, 4)
+    lat_bnds = lats_cor.reshape(nlat, nlon, 4)
+
+    # Add to the dataset
+    ds["lat_bnds"] = (
+        (ds[lat].dims[0], ds[lat].dims[1], "vertices"),
+        lat_bnds,
+    )
+    ds["lon_bnds"] = (
+        (ds[lon].dims[0], ds[lon].dims[1], "vertices"),
+        lon_bnds,
+    )
+
+    return ds
+
+
+def generate_bounds_rectilinear(ds, lat, lon):
+    """
+    Compute bounds for rectilinear grids.
+
+    The bounds will be attached as coords to the xarray.Dataset of the Grid object.
+    If no bounds can be created, a warning is issued. It is assumed but not ensured that no
+    duplicated cells are present in the grid.
```

Comment:
What is the difference between this and https://cf-xarray.readthedocs.io/en/latest/generated/xarray.Dataset.cf.add_bounds.html#xarray-dataset-cf-add-bounds ? Are there features here worth porting to cf-xarray ?